### ISSUE TITLE
RFC: Evaluate zipformer-arabic-v3.1 on testset and qdat

### ISSUE THREAD

**Issue by obadx** [obadx/prepare-quran-dataset#48]
## Status
Proposed · Target branch: `fastconformer`

## Objective
Evaluate `zipformer_p_arabic_v3.1.pt` (65.5M streaming Zipformer2 + CTC, 251-symbol alphabet, blank=250, no LM) on the same two suites `train_streaming.py` scores our own models on, each in **offline** (full utterance) and **streaming** (chunk-by-chunk cache-aware) modes:

- **Held-out muaalem test set** → report **`per_phonemes`**.
- **`obadx/qdat_bench`** → report **`per_phonemes` + the Tajweed (qdat column) metrics**.

Only the **`phonemes` level** is produced by this model (a single CTC head over one alphabet), so no sifat comparisons. Numbers go into the same `test_results_*.json` / `qdat_bench_test_results_*.json` files our baselines use, keeping them directly comparable to `w2v2bert-streaming-rnn` and `fastconformer-cache-aware`.

## The held-out test set (where it lives, how to use it)
The test split is selected by moshaf id. The ids are already defined in the training configs under `test_moshaf_ids` — e.g. `configs/train/streaming/train_config.yml`:

```yaml
test_moshaf_ids: ["19.0", "29.0", "30.0"]
```

Each id is a moshaf (reciter/reading) config of `obadx/muaalem-annotated-v3`; every ayah of that moshaf is an eval sample, phonetized with `quran-Transcript` at load time. (Any `configs/train/streaming/*.yml`, fastconformer branch included, carries the same `test_moshaf_ids`.)

The eval script must not hard-code ids: load the yml with `TrainConfig.from_yaml` and read `train_config.test_moshaf_ids`. The moshaf attribute metadata needed for phonetization comes from `obadx/mualem-recitations-annotated` (`moshaf_metadata` split), and a few segment-level recitation-way overrides come from `prepare_special_moshaf_ways`.

## How to evaluate correctly (functions to emulate)
All helpers referenced below are in `train_streaming.py` of **this** repo. Emulate their behaviour with the zipformer model; keep the same metric definitions so the results are comparable. (Line numbers are convenience hints only.)

**Test-set reference labels & PER:** use the exact label-prep + metric path our trainers use.
- `prepare_special_moshaf_ways` (≈`train_streaming.py:1018`) + `MoshafAttributes` from the `moshaf_metadata` split.
- `prepare_muaalem_dataset` (≈`:582`, with `is_testset=True`) builds+phonetizes the test samples.
- `prepare_dataset` (≈`:665`, with `is_testset=True`) filters to `DatasetDict{"test"}`.
- `DataCollatorCTCWithPadding` (≈`:891`) with `apply_augment=False` and `add_eos_token` from the config — call it on each batch to get reference `labels`.
- `IncrementalMetrics` (≈`:270`) with `SimpleNamespace(predictions=logits, label_ids=labels)`, passing `compute_result=True` on the last batch; report `per_phonemes` only.

**qdat_bench:** load `obadx/qdat_bench` (`split="train"`, cast audio `decode=False`), mirror the flag + filename cache pattern used in `__main__` (≈`:1767`).
- Offline: emulate `run_qdat_bench_test` (≈`:1047`) — per-sample forward, `ctc_decode` (≈`:212`), strip `multi_level_tokenizer.special_tokens`, write `qdat_bench_predictions_{suffix}.jsonl` with `levels_labels` and `level_to_scripts["phonemes"]`, then score with `eval_qdat_bench` (from `qdat_bench.eval_results`, ≈`:37`). Report `speech_metrics["per_phonemes"]` and `qdat_metrics`.
- Streaming: emulate `run_streaming_qdat_bench_test` (≈`:1178`) — same output/score, but with the chunked streaming loop.
- Optional: `compute_qdat_top_phonemes_per` (≈`:1390`) for a per-phoneme error ranking.
- `levels_labels` for the sifat levels must be present but may be empty lists (the zipformer emits a phonemes stream only) — `eval_qdat_bench` iterates every tokenizer level, so empty lists avoid KeyErrors; only `per_phonemes`/`qdat_metrics` are read.

**Phoneme-alphabet normalization (mandatory step):** the zipformer symbol table (251 entries) groups our base phoneme units (e.g. `قَ` → `ق`+`َ`, `ببَ` → `ب`+`ب`+`َ`, `اااااا` → six `ا`, `قڇ` → `ق`+`ڇ`). Every unit char already exists in our `vocab_streaming/vocab.json["phonemes"]`. Build an expansion map `tokens.txt symbol → unit-string`, assert no OOV (raise on any unit outside our vocab), then tokenize **both** predictions and references with `MultiLevelTokenizer` so id spaces match and PER is apples-to-apples with our baselines.

## New eval script — `zipformer_eval.py` (added in this repo)
Standalone CLI, no `Trainer`:

```
uv run python zipformer_eval.py \
  --checkpoint ./zipformer_p-arabic-v3/zipformer_p_arabic_v3.1.pt \
  --tokens ./zipformer_p-arabic-v3/tokens.txt \
  --config ./configs/train/streaming/train_config.yml \
  --out ./results-zipformer-v3.1 \
  [--offline] [--streaming] [--device] [--batch-size]
```

- **Model construction:** no `config.json` ships on the Hub, so the `Zipformer2` + `ctc_output` config is hard-coded and **validated against the state-dict shapes** (assert every key maps to a matching parameter). Start from the ~65.5M librispeech-default stack (`downsampling_factor=(1,2,4,8,4,2)`, `encoder_dim=(192,256,384,512,384,256)`, causal streaming); correct if shape asserts fail and confirm with Quran-Lab.
- **Features:** 80-dim kaldi fbank @16 kHz, povey window (`snip_edges=False`, `dither=0`) — **not** mel. Settings as in `quran_per_eval.py`.
- **Streaming:** chunked cache-aware decode via `get_init_states` + `streaming_forward` (trained at chunk sizes 8/16/24; v3.1 keeps the identical streaming interface).
- **Decoding:** greedy argmax, blank=250, ids → symbols via `tokens.txt`, then expansion to our phoneme ids (above).
- **Outputs** under `results-zipformer-v3.1/`, following our naming:
  - `test_results_zipformer_v3_1.json`, `test_results_streaming_zipformer_v3_1.json` — `per_phonemes`.
  - `qdat_bench_predictions_zipformer_v3_1[._streaming].jsonl`, `qdat_bench_test_results_zipformer_v3_1[._streaming].json` — `per_phonemes` + `qdat_metrics`.
  - `qdat_top_phonemes_per_zipformer_v3_1.json` (optional).

## About the model (icefall pointers)
- **Architecture:** `egs/librispeech/ASR/zipformer/zipformer.py` → `Zipformer2`; CTC head `ctc_output` (`nn.Sequential(Dropout, Linear, LogSoftmax)`); model wrapper + checkpoint loading in the same recipe's `pretrained.py` (`icefall.checkpoint.load_checkpoint` / `average_checkpoints`).
- **Streaming:** `export-onnx-streaming-ctc.py` in that recipe shows the chunked cache-aware loop (`get_init_states` → `streaming_forward` → `ctc_output`); onnx metadata: `T=61`, `decode_chunk_len=48`, `left_context_len`.
- **Mandatory symbol table:** decode raw ids with `tokens.txt` — ids are offset by one against `phoneme_units.json`, blank is 250 not 0; without it output is garbage.
- No LM anywhere by design; pure greedy CTC.

## Known caveats (model card — read before drawing conclusions)
1. **Chunk vs full context depends on domain:** in-domain, streaming chunks (16/24) match or beat full context; out-of-domain, full context is ~2 PER better. Our test set is in-domain; report both modes regardless.
2. **Free-choice madd (munfasil/aared) lengths are unreliable in the token stream** (labels fix 4 harakat; reciters choose 2/4/6). Duration grading should come from forced-alignment timing, not symbol run-length.
3. **The model transcribes the rule, not the deviation** on non-canonical recitation (e.g. full ikhfaa where annotators heard plain noon) — a measured limitation on the qdat quality-contrast attributes.
4. **Gated NPL-1.2 license** — HF auth required; never charge for the model or anything it powers.
5. **No `config.json` shipped** — the reconstructed architecture must be verified (state-dict shape asserts + Quran-Lab confirmation).

## Tests / acceptance
- **Smoke:** decode `assets/audio-sampels/test_sample_4.ogg` offline + streaming; assert non-empty phoneme strings and report the streaming-vs-offline matching ratio like `fastconformer_compare_streaming.py`. On clean in-domain audio the two paths should nearly agree.
- **Sanity:** qdat_bench `per_phonemes` lands in the same ballpark as `muaalem-model-v3_2` (model card reports comparable madd RMSE); wildly divergent numbers mean a feature/vocab bug, not model quality.
- **Tabulate** against existing `results-fastconformer-*` / `results-streaming-rnn-*` baselines (`aggregate_experiments.py` can fold them in).

### 1. Request access to the (gated) model
`zipformer_p_arabic_v3.1.pt` lives under the **Quran-Lab No-Profit License 1.2** and is gated: log in on HuggingFace and accept the access conditions on the model page. Then download the files you need (with your HF token available to `hf login` / `HUGGINGFACE_TOKEN`):

```bash
hf download Quran-Lab/zipformer_p-arabic-v3 \
  zipformer_p_arabic_v3.1.pt tokens.txt phoneme_units.json quran_per_eval.py \
  --local-dir ./zipformer_p-arabic-v3
```

`quran_per_eval.py` is the model owner's own evaluation loop: it shows the **exact streaming chunk loop and kaldi-fbank settings**. Use it as the reference for the feature/streaming details below.

### 2. Add icefall as an optional dependency (eval runs from this repo)
The `.pt` weights are an icefall *training checkpoint*; the architecture is icefall's `Zipformer2`. The eval script runs from this repository, so icefall is declared as an **optional dependency group `zipformer`** in `pyproject.toml`:

```toml
[project.optional-dependencies]
zipformer = ["icefall"]

[tool.uv.sources]
icefall = { git = "https://github.com/k2-fsa/icefall.git" }
```

Then (from this repo):

```bash
uv add icefall --optional zipformer     # adds the dep + git source above
uv sync --all-extras                    # installeverything incl. icefall
# or, to skip the other extras: uv sync --extra zipformer
uv run python -c "import icefall, torch, qdat_bench; print('ok')"
```

**Fallback** if icefall resists being installed from git (k2/kaldifeat build quirks): clone and sync it standalone, then wire it in editable:

```bash
git clone https://github.com/k2-fsa/icefall && cd icefall && uv sync --all-extras
cd streaming-train && uv add icefall --optional zipformer --editable ../icefall
```

## Definition of done
- [ ] Gated access granted; `.pt`, `tokens.txt`, `phoneme_units.json`, `quran_per_eval.py` downloaded
- [ ] `icefall` added as the `zipformer` extra (`uv add icefall --optional zipformer`); `import icefall` resolves in this repo's env
- [ ] `zipformer_eval.py` loads `zipformer_p_arabic_v3.1.pt` with state-dict shape assertions passing
- [ ] Expansion map from `tokens.txt` built with no OOV against our phonemes vocab
- [ ] `per_phonemes` produced for the test set and `per_phonemes` + Tajweed metrics for qdat_bench, offline and streaming
- [ ] Result JSONs saved under `results-zipformer-v3.1/` in our naming convention
- [ ] Smoke + sanity checks pass; run numbers tabulated against our baselines
